shark_turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -0,0 +1,68 @@
+from ..ops.wave_ops import Write, ExtractSlice, get_custom
+from .constraints import Constraint, HardwareConstraint
+from .._support.tracing import CapturedTrace, IndexingContext
+from .._support.indexing import IndexSymbol, IndexSequence
+from ..lang.global_symbols import *
+from .utils import simplify_index, get_mma_dimensional_mapping, get_hardware_vector_size
+import torch.fx as fx
+import numpy as np
+
+
+def get_vector_shape(
+    trace: CapturedTrace,
+    hardware_constraint: HardwareConstraint,
+    symbolic_shape: list[IndexSymbol],
+) -> list[int]:
+    mma_indices = get_mma_dimensional_mapping(trace)
+    return [
+        get_hardware_vector_size(dim, hardware_constraint, mma_indices)
+        for dim in symbolic_shape
+    ]
+
+
+def partition_strided_operators(trace: CapturedTrace, constraints: list[Constraint]):
+    """
+    This function analyzes the index sequence of operators in the graph
+    that are writes on 2d tensors. If the operator has an access pattern where
+    the strides are greater than one on a single dimension, this function splits the
+    operands into individual elements and constructs a write for
+    each individual element.
+    """
+
+    def has_strided_access(node: fx.Node) -> bool:
+        """
+        Checks for writes on 2d tensors with strided access on a single dimension.
+        """
+        custom = get_custom(node)
+        if isinstance(custom, Write) and len(custom.type.symbolic_shape) == 2:
+            strides = [simplify_index(custom.index[dim]).stride for dim in custom.index]
+            return sum(1 for stride in strides if stride > 1) == 1
+        return False
+
+    strided_operators = trace.walk(has_strided_access)
+    for operator in strided_operators:
+        custom = get_custom(operator)
+        simplified_index = {
+            dim: simplify_index(custom.index[dim]) for dim in custom.index
+        }
+        max_stride = int(max(simplified_index[dim].stride for dim in simplified_index))
+        hw_constraint = [c for c in constraints if isinstance(c, HardwareConstraint)][0]
+        shape = get_vector_shape(trace, hw_constraint, custom.type.symbolic_shape)
+        idxc = IndexingContext.current()
+        elements_per_thread = custom.elements_per_thread
+        if isinstance(elements_per_thread, IndexSymbol):
+            elements_per_thread = elements_per_thread.subs(idxc.subs)
+        with custom.graph.inserting_before(operator):
+            for i in range(elements_per_thread):
+                extract = ExtractSlice(custom.register_, [i], [1], [1]).add_to_graph(
+                    custom.graph
+                )
+                offset = np.unravel_index(i * max_stride, shape)
```

Comment:
What is the intuition behind delinearizing i * max_stride by the shape? not super obvious to me here

Reply:
Documentation from offline chat: it is i * max_stride because for example if we are writing 4 elements to a transposed matrix with fastest dim len of 16, then we'd need to write it at index 0[0 * stride], 16 [1 * stride], 32[2 * stride] , 48[3 * stride]